### PR TITLE
ITOS-32: Added error handlers & added queue and retry logic

### PR DESCRIPTION
**Pull request by dserkowski**
It is the continuation of this PR https://github.com/IsantePlus/openmrs-module-outgoing-exception/pull/5

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/outgoingmessageexceptions/api/retry/impl/PixRetryInvoker.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+			if (StringUtils.equals(outgoingMessage.getDestination(),
+					PixErrorHandlingService.SENDING_PATIENT_AFTER_PATIENT_CREATION_DESTINATION)) {
+				String ecid = mpiProvider.exportPatient(patient);
+				updatePatient(patient, ecid);
+			} else {
+				mpiProvider.updatePatient(patient);
+			}
+
+			outgoingMessage.setRetryReason("Retried successfully");
```
Comment (areklalo) [inline api/src/main/java/org/openmrs/module/outgoingmessageexceptions/api/retry/impl/PixRetryInvoker.java]: I think there you should use another field to contains information about successful retry

File: omod/src/main/resources/messages.properties [hunk 2]
```diff
@@ -16,16 +16,20 @@ ${project.parent.artifactId}.ccd.label=CCD Exceptions
...
+
 ${project.parent.artifactId}.details.messageNotFound=Message not found
 
 ${project.parent.artifactId}.message.header.timestamp=Timestamp
 ${project.parent.artifactId}.message.header.destination=Destination
-${project.parent.artifactId}.message.header.user=User
-${project.parent.artifactId}.message.header.failure=Success/Failure
\ No newline at end of file
+${project.parent.artifactId}.message.header.user=Owner
+${project.parent.artifactId}.message.header.failure=Successfully retried
```
Comment (areklalo) [inline omod/src/main/resources/messages.properties]: I think you need to change the key of this property

File: omod/src/main/resources/messages_es.properties [hunk 3]
```diff
@@ -27,5 +27,5 @@ ${project.parent.artifactId}.details.messageNotFound=Message not found
 
 ${project.parent.artifactId}.message.header.timestamp=Timestamp
 ${project.parent.artifactId}.message.header.destination=Destination
-${project.parent.artifactId}.message.header.user=User
-${project.parent.artifactId}.message.header.failure=Success/Failure
\ No newline at end of file
+${project.parent.artifactId}.message.header.user=Owner
+${project.parent.artifactId}.message.header.failure=Successfully retried
```
Comment (areklalo) [inline omod/src/main/resources/messages_es.properties]: Like above

File: omod/src/main/resources/messages_fr.properties [hunk 4]
```diff
@@ -14,18 +14,22 @@ ${project.parent.artifactId}.ccd.description=CCD Outgoing Exceptions
...
-${project.parent.artifactId}.message.header.timestamp=Timestamp
+${project.parent.artifactId}.details.messageNotFound=Message non trouvé
+
+${project.parent.artifactId}.message.header.timestamp=Horodatage
 ${project.parent.artifactId}.message.header.destination=Destination
-${project.parent.artifactId}.message.header.user=User
-${project.parent.artifactId}.message.header.failure=Success/Failure
\ No newline at end of file
+${project.parent.artifactId}.message.header.user=Propriétaire
+${project.parent.artifactId}.message.header.failure=Réessayé avec succès
```
Comment (areklalo) [inline omod/src/main/resources/messages_fr.properties]: Like above

**Review by areklalo — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/openmrs/module/outgoingmessageexceptions/OutgoingMessage.java [hunk 5]
```diff
@@ -259,7 +285,7 @@ public JsonElement serialize(OutgoingMessage src, Type typeOfSrc, JsonSerializat
 				JsonObject owner = new JsonObject();
 				
 				owner.addProperty("uuid", src.owner.getUuid());
-				owner.addProperty("name", src.owner.getUsername());
+				owner.addProperty("name", src.owner.getDisplayString());
```
Comment (pawelcieszko) [inline api/src/main/java/org/openmrs/module/outgoingmessageexceptions/OutgoingMessage.java:288]: why do you change it to DisplayString?

File: api/src/main/java/org/openmrs/module/outgoingmessageexceptions/api/OutgoingMessageExceptionsService.java [hunk 6]
```diff
@@ -63,9 +63,22 @@
...
+	
+	/**
+	 * Returns an serialized OutgoingMessage by id. It can be called by any authenticated user.
+	 * It is fetched in read only transaction.
+	 *
+	 * @param id
+	 * @return OutgoingMessage
+	 * @throws APIException
+	 */
 	@Authorized()
```
Comment (pawelcieszko) [inline api/src/main/java/org/openmrs/module/outgoingmessageexceptions/api/OutgoingMessageExceptionsService.java]: according to your changes the braces should be also removed

File: api/src/main/java/org/openmrs/module/outgoingmessageexceptions/api/retry/impl/PixRetryInvoker.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Service;
+
+@Service
+public class PixRetryInvoker implements RetryInvoker {
+
+	private static final Logger LOGGER = LoggerFactory.getLogger(PixRetryInvoker.class);
+
+	@Autowired
+    @Qualifier("adminService")
```
Comment (pawelcieszko) [inline api/src/main/java/org/openmrs/module/outgoingmessageexceptions/api/retry/impl/PixRetryInvoker.java]: fix identation

File: omod/src/main/java/org/openmrs/module/outgoingmessageexceptions/web/controller/rest/OutgoingMessageExceptionsRestController.java [hunk 8]
```diff
@@ -78,7 +78,7 @@ public String getMessagesList(@RequestParam(value = "pageIndex", defaultValue =
 	public String getMessageById(@PathVariable Integer id) throws JsonProcessingException {
 		if (Context.hasPrivilege(OutgoingMessageExceptionsConstants.GET_OUTGOING)) {
 			logger.debug("Get Single message reached by message id");
```
Comment (pawelcieszko) [inline omod/src/main/java/org/openmrs/module/outgoingmessageexceptions/web/controller/rest/OutgoingMessageExceptionsRestController.java]: "Single" lowercase

File: omod/src/main/resources/messages.properties (hunk 2 above)
Comment (pawelcieszko) [reply omod/src/main/resources/messages.properties]: and add a new line at the eof

File: omod/src/main/webapp/fragments/exceptionDetails.gsp [hunk 9]
```diff
@@ -21,7 +21,7 @@
             <td>${ outgoingMessage.timestamp }</td>
         </tr><tr>
             <th class="label">${ ui.message(artifactId + ".details.failureReason.label") }</th>
-            <td>${ outgoingMessage.failureReason }</td>
+            <td><textarea style="width:100%" disabled>${ outgoingMessage.failureReason }</textarea></td>
```
Comment (pawelcieszko) [inline omod/src/main/webapp/fragments/exceptionDetails.gsp]: Are you sure that using disabled text area as a failure reason container is a good idea?

**Review by pawelcieszko — CHANGES_REQUESTED** [review]

File: omod/src/main/webapp/fragments/exceptionDetails.gsp (hunk 9 above)
Comment (dserkowski) [reply omod/src/main/webapp/fragments/exceptionDetails.gsp]: I will change it to readonly. Readonly doesn't change colours of the field

File: api/src/main/java/org/openmrs/module/outgoingmessageexceptions/OutgoingMessage.java (hunk 5 above)
Comment (dserkowski) [reply api/src/main/java/org/openmrs/module/outgoingmessageexceptions/OutgoingMessage.java:288]: It provides more information. Eg. user admin has the username field blank, but name generated by DisplayString is "Super User (null)"

Comment (areklalo) [reply api/src/main/java/org/openmrs/module/outgoingmessageexceptions/OutgoingMessage.java:288]: Ok.

**Pull request merged**